Add string.split implementation

Validark opened the pull request:
- Adds string.split and tests. Modeled after the file which adds Roblox's `math.clamp` function.

tiffany352 commented:
Here are our internal unit tests for this function, reformatted to the unit testing framework used here.

```lua
local char = string.char
local ZWJ = char(0xe2, 0x80, 0x8d)

assert.are.same({ "" }, string.split("", ","))
assert.are.same({ "foo", "", "bar" }, string.split("foo,,bar", ","))
assert.are.same({ "", "foo" }, string.split(",foo", ","))
assert.are.same({ "foo", "" }, string.split("foo,", ","))
assert.are.same({ "", "" }, string.split(",", ","))
assert.are.same({ "", "", "" }, string.split(",,", ","))
assert.are.same({ "" }, string.split("", "~~~"))
assert.are.same({ "~~" }, string.split("~~", "~~~"))
assert.are.same({ "~~ ~~" }, string.split("~~ ~~", "~~~"))
assert.are.same({ "foo", "bar" }, string.split("foo~~~bar", "~~~"))
assert.are.same({ "foo", "", "bar" }, string.split("foo~~~~~~bar", "~~~"))
assert.are.same({ "", "foo" }, string.split("~~~foo", "~~~"))
assert.are.same({ "foo", "" }, string.split("foo~~~", "~~~"))
assert.are.same({ "", "" }, string.split("~~~", "~~~"))
assert.are.same({ "", "", "" }, string.split("~~~~~~", "~~~"))
assert.are.same({ "", "", "O" }, string.split("OOOOO", "OO"))
assert.are.same({ "   ws   " }, string.split("   ws   ", ","))
assert.are.same({ "foo ", " bar" }, string.split("foo , bar", ","))
assert.are.same({ "我很高兴", "你呢？" }, string.split("我很高兴，你呢？", "，"))
assert.are.same({ "👩", "👩", "👧", "👧" }, string.split("👩‍👩‍👧‍👧", ZWJ))
assert.are.same({ "foo", "bar" }, string.split("foo\0bar", "\0"))
assert.are.same({ "foo", "bar", "" }, string.split("foo\0bar\0", "\0"))
assert.are.same({ "foo", "bar" }, string.split("foo\0\0bar", "\0\0"))
assert.are.same({ "foo\0" }, string.split("foo\0", "\0\0"))
assert.are.same({ "foo", "\0" }, string.split("foo\0\0\0", "\0\0"))
assert.are.same({ }, string.split("", ""))
assert.are.same({ "a", "b", "c" }, string.split("abc", ""))
assert.are.same({ char(0xef), char(0xbc), char(0x9f) }, string.split("？", ""))
```

Validark commented:
Thank you so much for providing this! I added it in. I think it is pretty funny you can use `string.split` to split up a family. Luckily, a call to `table.concat` should bring them back together.

LPGhatguy commented:
If it passes our internal tests, then it looks good to me!
